### PR TITLE
ASoC: intel: sof_sdw: init all aggregated codecs

### LINKED ISSUE #1953: [BUG][CML]Aplay0,2 rt1308 right speaker have sound output when muted speaker in alsamixer.
**Describe the bug**
Aplay0,2 rt1308 right sepaker have sound output when muted speaker in alsamixer

**To Reproduce**
1.Boot up system successfully.
2.Switch speaker to mute in alsamixer.
2.Aplay -Dhw:0,2 -fs16_le -c2 -r 48000 -vv -i 00.wav

**Reproduce rate**
5/5

**Expected result**
Speaker should't have sound output when selected headphone in soundsetting.

**Actual result**
Aplay0,2 rt1308 right speaker have sound output when muted speaker in alsamixer.

**Test recipe**
kernel: topic/sof-dev commit a272777 + #1943 +#1897
FW: cml-011-drop-stable commit: 37ad771
tplg: cml-011-drop-stable + 827d1f6 which is the same as https://github.com/bardliao/sof/tree/cml-topology
platform: CML-H Laptop with ALC711+ ALC1308+ ALC715 in SDW mode

[dmesg.log](https://github.com/thesofproject/linux/files/4397130/dmesg.log)
[sof_error_trace28.log](https://github.com/thesofproject/linux/files/4397131/sof_error_trace28.log)
[sof_trace.log](https://github.com/thesofproject/linux/files/4397132/sof_trace.log)

### PR DESCRIPTION
**Pull request by bardliao**
Init codecs which belong to the same group id on all links.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

Fixes #1953

### CODE REVIEW

**Comment by bardliao** [comment]
> The code looks fine @bardliao but I don't know how things worked before? Is this required now that we have the NO_AGGREGATION removed?

Yes, NO_AGGREGATION flag will hide the issue.

**Comment by RanderWang** [comment]
> The code looks fine @bardliao but I don't know how things worked before? Is this required now that we have the NO_AGGREGATION removed?

yes, without aggregation we only need to init codec on current link for current BE DAI.

File: sound/soc/intel/boards/sof_sdw.c [hunk 1]
```diff
@@ -411,25 +411,30 @@ static int create_codec_dai_name(struct device *dev,
...
+			codec_index = find_codec_info_part(part_id);
 
-		if (codec_info_list[codec_index].init)
-			codec_info_list[codec_index].init(link, dai_links,
-						 &codec_info_list[codec_index],
-						 playback);
-	}
+			if (codec_index < 0)
+				return codec_index;
+			if (link->adr_d[i].endpoints->group_id != group_id)
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c:434]: @bardliao dont you have to check if the enpoint is aggregated here as well?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:434]: > @bardliao dont you have to check if the enpoint is aggregated here as well?

group_id starts with 1 when it is aggregated. In other words, the aggregated flag is always 0 when group_id is 0. Also, ``while (link->mask && group_id)`` will be invalid if a codec is aggregated with group_id 0. In my opinion, we should fix it in the match table if it sets an non zero group id to a non aggregated endpoint or a zero group id to an aggregated endpoint.

Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c:434]: @bardliao but you have a do..while{} loop, so the first iteration could be even with group_id == 0 no?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:434]: > @bardliao but you have a do..while{} loop, so the first iteration could be even with group_id == 0 no?

Yes, we should do codec init in the first iteration no matter it is aggregated or not.

Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c:434]: ahh now I get it. The commit message confused me a bit. I think you should just say: "Init codecs which belong to the same group id on all links".

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:434]: > ahh now I get it. The commit message confused me a bit. I think you should just say: "Init codecs which belong to the same group id on all links".

Done.

File: sound/soc/intel/boards/sof_sdw.c [hunk 2]
```diff
@@ -411,25 +411,30 @@ static int create_codec_dai_name(struct device *dev,
...
 {
 	int i;
 
-	for (i = 0; i < link->num_adr; i++) {
-		unsigned int part_id;
-		int codec_index;
-
-		part_id = SDW_PART_ID(link->adr_d[i].adr);
-		codec_index = find_codec_info_part(part_id);
+	do {
```
Comment (lyakh) [inline sound/soc/intel/boards/sof_sdw.c:418]: Took me a while to figure out why we need to loop here. It looks like now with aggregation the loop over links in `sof_card_dai_links_create()` only calls `create_sdw_dailink()` for the first link in each group. That's why you now have to walk over all links, looking for others from the same group. Maybe you could add a comment about that for the next reader :-)

File: sound/soc/intel/boards/sof_sdw.c [hunk 3]
```diff
@@ -411,25 +411,30 @@ static int create_codec_dai_name(struct device *dev,
...
+			if (link->adr_d[i].endpoints->group_id != group_id)
+				continue;
+			if (codec_info_list[codec_index].init)
+				codec_info_list[codec_index].init(link,
+						dai_links,
+						&codec_info_list[codec_index],
+						playback);
+		}
+		link++;
+	} while (link->mask && group_id);
```
Comment (kv2019i) [inline sound/soc/intel/boards/sof_sdw.c:443]: @bardliao I get the logic by looking at the conversation about you had with @ranj063 but based on just the git commit message and code, this check for "while(group_id)" still seems cryptic. Maybe add a comment stating that group_id>0 means codec is aggregated

**Comment by plbossart** [comment]
@bardliao please add comments to explain that the loop is done once and group_id > 0 means aggregation. You and I know this, but others may not.

File: sound/soc/intel/boards/sof_sdw.c (hunk 3 above)
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:443]: Done

Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c:443]: @bardliao a clearer comment would be something that says something like: "The loop initializes the codec if it does not belong to an aggregated group or initializes the first codec in an aggregated group and iterates through the links to initialize the rest of the codecs in the same group"

Comment (lyakh) [reply sound/soc/intel/boards/sof_sdw.c:443]: Agree, let's improve the comment a bit. In fact this whole aggregation implementation seems a bit inconvenient to handle. I understand that it has been added as an after-thought. So what we have now is the outer loop over links in `sof_card_dai_links_create()`, which calls `create_sdw_dailink()` which then continues looping over the same list of links, looking for groups. Then `set_codec_init_func()` is called which now again loops over the same links... It isn't anything exceptional, but maybe someone comes up with a good idea to simplify this a bit.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:443]: Thanks @ranj063 @lyakh The comment is updated with slight modification from @ranj063 's suggestion. A link may contain more than one codec and we will go through all of them in the first iteration. An assumption here is all codecs in the same link will either not aggregated or with in the same aggregated group. Note that the adr link here doesn't mean the physical link. A physical link can be separate to different adr links.

Comment (kv2019i) [reply sound/soc/intel/boards/sof_sdw.c:443]: Thanks @bardliao -- I'm afraid the comment is not helping enough yet. Even with @ranj063 's changes, there's still no explanation of the special meaning of group_id==0. How about:

"Initialize the codec. If codec is part of an aggregated group (group_id>0), initialize all codecs belonging to same group."

?

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:443]: > Thanks @bardliao -- I'm afraid the comment is not helping enough yet. Even with @ranj063 's changes, there's still no explanation of the special meaning of group_id==0. How about:
> 
> "Initialize the codec. If codec is part of an aggregated group (group_id>0), initialize all codecs belonging to same group."
> 
> ?

Thanks @kv2019i That is clear and simple.

**Comment by kv2019i** [comment]
Jenkins build stalled, CI errors are known or unrelated (GLK sof-logger error), merging.

**Pull request merged**